src/fairchem/core/trainers/base_trainer.py

```diff
@@ -356,17 +361,128 @@ def load_datasets(self) -> None:
+                        for lr_target, references in otf_elementrefs[0].items():
+                            path = save_checkpoint(
+                                references.state_dict(),
+                                self.config["cmd"]["checkpoint_dir"],
+                                f"{target}_linref.pt",
+                            )
+                            logging.info(
+                                f"{lr_target} linear references have been saved to: {path}"
+                            )
+
+                distutils.broadcast_object_list(otf_elementrefs, src=0)
+                # make sure all of the element reference modules are on the same device
+                self.elementrefs.update(otf_elementrefs[0])
+                # set config so that references are not refit
+                self.config["dataset"]["transforms"]["element_references"]["fit"][
+                    "fitted"
+                ] = True
+            else:  # load pre-fitted linear references from file
+                self.elementrefs[target] = create_element_references(
+                    file=elementrefs[target].get("file"),
+                )
+
+        # load or fit normalizers for the dataset.
+        normalizers = self.config["dataset"].get("transforms", {}).get("normalizer", {})
         self.normalizers = {}
-        if normalizer:
-            for target in normalizer:
-                self.normalizers[target] = Normalizer(
-                    mean=normalizer[target].get("mean", 0),
-                    std=normalizer[target].get("stdev", 1),
+        for target in normalizers:
+            if target == "fit" and not normalizers["fit"].get("fitted", False):
+                otf_normalizers = [
+                    {target: None for target in normalizers["fit"]["targets"]}
+                ]
+                # only carry out the fit on master and then broadcast
+                if distutils.is_master():
+                    otf_normalizers = [
+                        fit_normalizers(
+                            targets=normalizers["fit"]["targets"],
+                            element_references=self.elementrefs,
+                            dataset=self.train_dataset,
+                            batch_size=normalizers["fit"].get(
+                                "batch_size", self.config["optim"]["batch_size"]
+                            ),
+                            num_batches=normalizers["fit"].get("num_batches"),
+                            num_workers=self.config["optim"]["num_workers"],
+                            seed=self.config["cmd"]["seed"],
+                        )
+                    ]
+                    # save the normalization for possible subsequent use
+                    if not self.is_debug:
+                        save_checkpoint(
+                            otf_normalizers[0],
+                            self.config["cmd"]["checkpoint_dir"],
+                            "normalizers.pt",
```

Comment:
For elemental references you save out files on a per target basis. Here you're doing it for all of them in a single file. Any particular reason you're doing one over the other?

Reply:
Only because references will realistically only be on energy. Though perhaps we should follow the same logic of a single file for the case where someone wants more than one energy head?

Reply:
Set them to follow the same logic now, this makes the code cleaner :)